Digest option 2.5

LPardue opened the pull request:
This PR attempts to capture my interpretation of the discussion during the HTTP Interim meeting Feb 2022. Editorial judgement has been applied with judicious KISS.

Executive summary:

1. Obsolete 3230.
  *  The suggestion was to put this in an appendix. However, if we are obsoleting Digest, I don't see much benefit in talking through many details. Stop using it and use the substitutes we spend lots of text on. The text ended up as half a page and I kept it in the Introductory section 1
2. Having a mix of sf-dictionary and sf-list that need to reference each other is really annoying. I'm sure there's a way to square the circle about using List items as lookups for Dictionary keys but after already spending hours noodling on it, it seems like an accident waiting to happen.
  * So I cheated. I switched Want-Representation-Digest and Want-Content-Digest to be Dictionaries too. It's just that the member value is an integer between 0 and 10 that expresses a preference weight. (Thanks MT for the suggestion to use int)
  * Using a dictionary here disposes of the 'q' parameter qvalue problem I mentioned in the meeting
3. New IANA table for the new headers
  * This is another slight cheat that lets us avoid updating a table that implementers of an obsolete RFC might still rely on.
  * It imports all the old headers from before, including insecure stuff that people previously said they wanted.

What this doesn't do is bikeshed on the header name. I suggest we try to find consensus on the general changes in the draft and follow up with issues for any bikeshed issues.

reschke commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -1,5 +1,5 @@
 ---
-title: Digest Fields
+title: Structured Digest Fields
```
I would keep that out of the title; it's a syntax detail.

LPardue replied:
Fine by me!

tfpauly replied:
Agreed, doesn't need to be in the title.

ioggstream replied:
```suggestion
title: Digest Fields
```

tfpauly commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+integrity of a resource the was reconstructed from parts retrieved using
+multiple requests or connections.
+
+This document obsoletes RFC 3230 and therefore the Digest and Wants-Digest HTTP
```
```suggestion
This document obsoletes RFC 3230 and therefore the Digest and Want-Digest HTTP
```

martinthomson replied:
This is a repeat of before, so

```suggestion
This document obsoletes RFC 3230 and the Digest and Want-Digest HTTP
```

tfpauly commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -522,16 +492,16 @@ to the target resource (see {{Section 2 of PATCH}}).
 # Security Considerations
 
-## Digest Does Not Protect the Full HTTP Message
+## HTTP Messages are not protected in full
```
Use title case here, rather than sentence case?

ioggstream replied:
```suggestion
## HTTP Messages Are Not Protected In Full
```

tfpauly commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -648,69 +606,50 @@ validation of the algorithm types, number of validations, or the size of content
-This memo sets this specification to be the establishing document for the [HTTP Digest
-Algorithm
-Values](https://www.iana.org/assignments/http-dig-alg/) registry.
+This memo sets this specification to be the establishing document for the [HTTP Structured Digest
```
Should we also formally close the old registry?

LPardue replied:
Thanks, you've highlighted an issue with section I'd missed. 

Other options tried to update the old registry. That's not the case here, I'm proposing a new registry and I need to update the text that does that to read better.

I don't know what the formalities for closing a registry are in relation to an obsoleted header. Does that mean it vanishes from IANA (seems bad), or that it can no longer have new entries added to it, or something else?

ioggstream replied:
@mnot any hint for this?

LPardue replied:
Paging @mnot again to test his email filters

mnot replied:
Ah, it's `mention@`, not `mentions` -- try again?

LPardue replied:
@mnot please answer the question 😉

mnot replied:
It works!

Oh, the *question*. Right. I don't know that anything needs to happen to the registry, unless you want to close it down (i.e., not allow updates). If that's the case, you should say so in IANA considerations. Generally a lot of this gets negotiated tweaked later in the process when IANA does their review.

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+Representation-Digest field can be used for the integrity of HTTP
+representations. The Content-Digest field can be used for the integrity of
+HTTP message content. Want-Representation-Digest and Want-Content-Digest can be
+used to indicate a sender's interest and preferences for receiving respective integrity fields.
```
```suggestion
used to indicate a sender's interest and preferences for receiving the respective
integrity fields.
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+used to indicate a sender's interest and preferences for receiving respective integrity fields.
 
-This document obsoletes RFC 3230.
+This document obsoletes RFC 3230 and therefore the Digest and Wants-Digest HTTP
```
```suggestion
This document obsoletes RFC 3230 and the Digest and Want-Digest HTTP
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
-Want-Digest and Want-Content-Digest can be used to indicate a sender's desire to
-receive integrity fields respectively.
+This document defines HTTP fields that support integrity checksums. The
+Representation-Digest field can be used for the integrity of HTTP
+representations. The Content-Digest field can be used for the integrity of
+HTTP message content. Want-Representation-Digest and Want-Content-Digest can be
```
Later you put content integrity first with representation integrity second.  Here, it's the other way around.  Maybe try to make that consistent.  (I'd prefer content-first ordering if it came down to a coin toss.)

LPardue replied:
Ah I know what's happened here. I started on the path of putting content digest in my local copy. However, I soon realized that to keep it consistent I would need to propogate changes through the document, which was very disruptive and not critical to this proposal. I thought I unwound all those changes, I missed this bit, sorry. 

I think the ordering of concepts should be tracked in a seperate ticket, WDYT?

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
-This document obsoletes [RFC3230].
+Second, representation data integrity, which acts on representation data ({{Section 3.2
+of SEMANTICS}}). This supports more advanced use cases such as validating the
+integrity of a resource the was reconstructed from parts retrieved using
```
```suggestion
integrity of a resource that was reconstructed from parts retrieved using
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
-this document introduces the `Content-Digest` request and response header and trailer field; see
+The HTTP fields defined in this document can be used for HTTP integrity. Senders
+choose a hashing algorithm and calculate a checksum from an input related to the
+HTTP message, the algorithm identifier and checksum are transmitted in an HTTP
```
```suggestion
HTTP message. The algorithm identifier and digest are transmitted in an HTTP
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
-To support use-cases where a simple checksum of the content bytes is required,
-this document introduces the `Content-Digest` request and response header and trailer field; see
+The HTTP fields defined in this document can be used for HTTP integrity. Senders
+choose a hashing algorithm and calculate a checksum from an input related to the
```
```suggestion
choose a hashing algorithm and calculate a digest from an input related to the
```

"checksum" has a particular meaning, where I *think* digest is more generic.

ioggstream replied:
@martinthomson is `hash` ok?

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+The HTTP fields defined in this document can be used for HTTP integrity. Senders
+choose a hashing algorithm and calculate a checksum from an input related to the
+HTTP message, the algorithm identifier and checksum are transmitted in an HTTP
+field. Receivers can validate the checksum for integrity purposes. Hashing
```
```suggestion
field. Receivers can validate the digest for integrity purposes. Digest
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+algorithms are registered in the HTTP Digest Algorithm Values Registry (see
+{{algorithms}}).
+
+Selecting the data on which checksums are calculated depends on the use case of
```
I'll stop here, but here's another one.

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+HTTP messages. This document provides different headers for HTTP representation
+data and HTTP content.
+
+This document defines the `Representation-Digest` request and response header
+and trailer field ({{representation-digest}}) that contains a checksum value
```
```suggestion
This document defines the `Representation-Digest` field ({{representation-digest}}) that contains a value
```

LPardue replied:
That's a decent suggestion, let's track it in https://github.com/httpwg/http-extensions/issues/1941

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+Since `Content-Digest` is a Dictionary, it can contain multiple
+members. This could be used, for example, to attach multiple checksums
+calculated using different hashing algorithms in order to support a population
+of endpoints with different evolving capabilities. Such an approach could
```
```suggestion
of endpoints with different or evolving capabilities. Such an approach could
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
+Since `Representation-Digest` is a Dictionary, it can contain multiple
+members. This could be used, for example, to attach multiple checksums
+calculated using different hashing algorithms in order to support a population
+of endpoints with different evolving capabilities. Such an approach could
```
```suggestion
of endpoints with different or evolving capabilities. Such an approach could
```

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,413 @@ normative:
   {: vspace="0"}
-  sha-512
-  : * Digest Algorithm: sha-512
+  SHA-512
```
Did you really have to all-caps these?  sf-dictionary keys are lcalpha.

Also, sf-dictionary keys, as lcalpha, don't allow hyphens...

LPardue replied:
Honest truth, I never liked the style of table initialization that digest used. You'll see that in the old text we have a list of "things" that then define the values for columns in https://www.iana.org/assignments/http-dig-alg/http-dig-alg.xhtml. Is the "thing" the algorithm (SHA-512) or is it the token (sha-512), it's ambiguous.

In 
What's in here is a hangover from proposal two which needed several columns to address the old and new header algo identifer. Because of the above ambiguity, I tried to change things. But it's not really needed here. I think we should just switch to a table format and it would make things clearer.

martinthomson replied:
Oh, I see what you were doing here.  I don't much mind.  I guess it's good because the change meant that I noticed the hyphen issue, which is a real problem.  The formatting I will leave to you.

LPardue replied:
yeah I clearly messed up the algo key thing, I'll get that fixed up thank you

jricher replied:
Just a note, sf-dictionary keys :do: allow hyphens, but not as the first character:

```
key           = ( lcalpha / "*" )
                *( lcalpha / DIGIT / "_" / "-" / "." / "*" )
```

LPardue replied:
hehe. So I guess we can stick with `sha-nnn` or drop the hypen. I don't really are much either way.

martinthomson replied:
Ah. Saved. I misread the text.

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -559,20 +529,22 @@ metadata are discussed in {{usage-in-signatures}}.
-insecure and collision-prone algorithms do not provide adequate guarantees
-(see {{algorithms}}).
+Digest fields SHOULD always be used over a connection that provides integrity at
+the transport layer that protects HTTP fields.
```
```suggestion
the transport layer to protect HTTP fields from modification.
```

Separately, this could be read to imply that a TCP checksum is sufficient.  I'm not enthusiastic about that, but...ok.

LPardue replied:
Your comment applies to the current version of the document too. we were kind of trying to lump TCP and TLS (transport layer security) together here. Maybe we can rephrase things but I would like to track that independendently of the proposal wholesale.

LPardue replied:
tracked in #1940

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -559,20 +529,22 @@ metadata are discussed in {{usage-in-signatures}}.
+Digest fields SHOULD always be used over a connection that provides integrity at
+the transport layer that protects HTTP fields.
+
+A `Representation-Digest` field using NOT RECOMMENDED hashing algorithms SHOULD NOT be used in
```
```suggestion
A `Representation-Digest` field using insecure hashing algorithms MUST NOT be used in
```

LPardue replied:
This seems out of scope for this proposal. Unless you really mean to suggest that for new Digest fields we should say MUST NOT. But that goes agaisnt the design we agreed where people that want to use MD5 can. I made  https://github.com/httpwg/http-extensions/issues/1929 to track this seperately.

martinthomson reviewed (changes requested):
I think that you need to think very carefully about your use of "digest", "integrity", "hash", and "checksum" so that you are sure that you know what each means.  I started to try to clean this up, but it looks to be a more substantial problem than suggestions can address.

The main reason I think you need changes though is that your keys don't fit the sf-dictionary syntax.

LPardue commented:
Thanks for the review comment Martin!

> I think that you need to think very carefully about your use of "digest", "integrity", "hash", and "checksum" so that you are sure that you know what each means. I started to try to clean this up, but it looks to be a more substantial problem than suggestions can address.

This proposal is branch of of proposal two, which included the `Digest` field too (and mentioned it a lot!). So I was attempting to disambiguate between the field and the product of doing some hashing. I could have failed in that by overlooking the nuances of the terms. However, now I think about it more, in this proposal we don't mention the `Digest` much, so switching back to digest is probably ok. I'll do that in a single commit. Thanks.

> The main reason I think you need changes though is that your keys don't fit the sf-dictionary syntax.

Quite right. We need to pick identifier that fit, that's straightforward and I'll follow up.

ioggstream commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,414 @@ normative:
-- {{digest}} describes concepts related to representation digests. It also
-  defines the Digest request and response header and trailer field,
+- {{representation-digest}} defines the Representation-Digest request and response header and trailer field,
 - {{content-digest}} defines the Content-Digest request and response header and trailer field,
-- {{want-fields}} defines the Want-Digest and Want-Content-Digest request and response header and
+- {{want-fields}} defines the Want-Representation-Digest and Want-Content-Digest request and response header and
   trailer field,
-- {{algorithms}} describes algorithms and their relation to Digest,
+- {{algorithms}} describes algorithms and their relation to the fields defined in this document,
 - {{state-changing-requests}} details computing representation digests,
 - {{examples-unsolicited}} and {{examples-solicited}} provide examples of using
-  Digest and Want-Digest.
+  Representation-Digest and Want-Representation-Digest.
```
@LPardue do we need it?

LPardue replied:
I'd keep it for now at least.

LPardue commented:
I think I've addressed the review comments made so far, or spun them out is separate follow up work.

Are there any more comments on the *design details* in this proposal? Summarizing

* New headers are SF Dictionary
* Want-* headers express preference using a 'q' parameter, which is an integer in the range 0 through 10.

mnot commented:
FWIW - see
  https://httpwg.org/http-extensions/structured-digest2-and-content-digest-appendix-rfc3230/draft-ietf-httpbis-digest-headers.html
for what this would look like in situ

ioggstream commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,372 @@ normative:
+choose a hashing algorithm and calculate a digest from an input related to the
+HTTP message, the algorithm identifier and digest are transmitted in an HTTP
+field. Receivers can validate the digest for integrity purposes. Hashing
+algorithms are registered in the HTTP Digest Algorithm Values Registry (see
```
```suggestion
algorithms are registered in the "Hash Algorithms for HTTP Digest Fields" (see
```

ioggstream commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,372 @@ normative:
-Digest: sha-512=WZDPaVn/7XgHaAy8pmojAkGWoRx2UFChF41A2svX+TaPm
-                AbwAgBWnrIiYllu7BNNyealdVLvRwE\nmTHWXvJwew==
+Representation-Digest: sha-512=:WZDPaVn/7XgHaAy8pmojAkGWoRx2UFChF41A2svX+TaPm
+                AbwAgBWnrIiYllu7BNNyealdVLvRwE\nmTHWXvJwew==:
```
```suggestion
                AbwAgBWnrIiYllu7BNNyealdVLvRwEmTHWXvJwew==:
```
typo

ioggstream replied:
:shrug:

ioggstream commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,372 @@ normative:
+| unixsum        | insecure | The algorithm used by the UNIX "sum" command. | [RFC4648], [RFC6234], [UNIX], this document. |
+| unixcksum      | insecure | The algorithm used by the UNIX "cksum" command. | [RFC4648], [RFC6234], [UNIX], this document. |
+| adler          | insecure | The ADLER32 algorithm.                          | [RFC1950], this document. |
+| crc32          | insecure | The CRC32c algorithm.                           | {{!RFC4960}} appendix B, this document. |
```
```suggestion
| crc32c         | insecure | The CRC32c algorithm.                           | {{!RFC4960}} appendix B, this document. |
```

LPardue replied:
I agree with the change of algorithm name, but I'll hold off merging to avoid the conversation being resolved and hard to find.

ioggstream commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -602,44 +533,30 @@ a proof of integrity "at rest" unless the whole (e.g. encoded) content is persis
+to choose hashing algorithms from the IANA Hash Algorithms for HTTP Digest Fields registry; see
+{{establish-hash-algorithm-registry}}.
 
 To help endpoints distinguish weaker algorithms from stronger ones,
 this document adds to the IANA Digest Algorithm Values registry
```
```suggestion
```

The old table is not updated anymore

ioggstream commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -1022,29 +907,29 @@ HTTP/1.1 200 OK
+Representation-Digest: sha-256=:4REjxQ4yrqUVicfSKYNO/cF9zNj5ANbzgDZt3/h3Qxo=:,
+        sha-512=:pxo7aYzcGI88pnDnoSmAnaOEVys0MABhgvHY9+VI+ElE6
+                0jBCwnMPyA/s3NF3ZO5oIWA7lf8ukk+\n5KJzm3p5og==:
 
 iwiAeyJoZWxsbyI6ICJ3b3JsZCJ9Aw==
```
```suggestion

iwiAeyJoZWxsbyI6ICJ3b3JsZCJ9Aw==
```

typo

LPardue replied:
what is the typo here? it looks identical to me

ioggstream replied:
Yes, my mistake.

ioggstream reviewed (changes requested):
Thanks @LPardue for your efforts, and to the community for the review!

1. I'm really in favor of an appendix on Digest and I'll take care of that.
2. :clap: on dictionaries for want-*. I think we should identify a common way to migrate qvalue  for all sf- 
3. Hash Digest Table: to import old numeric algos like CRC32C we need to specify how ot encode them in sf-binary
4. ok to defer bikesheds in another PR

LPardue commented:
> to import old numeric algos like CRC32C we need to specify how ot encode them in sf-binary

I don't know that we do. The output of the algorithms seems to vary between 16 bits and 32 bits. So in other words a byte sequence of 2 to 4 bytes. That's trivial to encode in accordance with  https://www.rfc-editor.org/rfc/rfc8941.html#name-byte-sequences

ioggstream commented:
> in other words a byte sequence of 2 to 4 bytes. 
> That's trivial to encode in accordance with https://www.rfc-editor.org/rfc/rfc8941.html#name-byte-sequences

You are right, ignore my consideration above :)

LPardue commented:
In https://github.com/httpwg/http-extensions/pull/1919/commits/008d975166b52956e254c5ae30b50342ab8ebb9c I've invoked our SF linter and fixed all it' warning alongide the artwork width nits. 

Because the field names are long in this proposal, there's a lot of wrapping needed. In follow up bikeshed, we can probably tidy that up for most things except the sha-512 examples.

ioggstream reviewed (approved):
It's ok to me once added a Digest appendix.

I can take care on that in a separate PR.

martinthomson commented on draft-ietf-httpbis-digest-headers.md:
```diff
@@ -77,442 +76,385 @@ normative:
+{{algorithms}}).
+
+Selecting the data on which digests are calculated depends on the use case of
+HTTP messages. This document provides different headers for HTTP representation
+data and HTTP content.
```
These concepts really need to be defined.  Probably right here.

ioggstream replied:
@martinthomson these concepts are defined in SEMANTICS and referenced in Notational Convention. There's an Annex explaining this stuff too. Am I missing something?

To ease this kind of work, I filed an issue on kramdown-rfc2629 that could help simplifying referencing concepts when using markdown. https://github.com/cabo/kramdown-rfc2629/issues/155

martinthomson replied:
I was looking for a citation right here.  It seemed like the right place for it.  Given how critical they are, it seems prudent to cite the relevant sections of {{SEMANTICS}} right here.

LPardue replied:
The HTML makes this easier to reason about. In https://httpwg.org/http-extensions/draft-ietf-httpbis-digest-headers.html#name-introduction. We cite the relevant sections in the next paragraphs. Part of the editorial work on reordering sections will refactoring things a little and sort out citing at the earliest and kost convenient time
